presto-main/src/main/java/com/facebook/presto/memory/ClusterMemoryManager.java

```diff
@@ -107,7 +121,63 @@ public void process(Iterable<QueryExecution> queries)
         return ImmutableMap.copyOf(pools);
     }
 
-    private void updateNodes()
+    private MemoryPoolAssignmentsRequest updateAssignments(Iterable<QueryExecution> queries)
+    {
+        ClusterMemoryPool reservedPool = pools.get(RESERVED_POOL);
+        ClusterMemoryPool generalPool = pools.get(GENERAL_POOL);
+        long version = memoryPoolAssignmentsVersion.incrementAndGet();
+        // Check that all previous assignments have propagated to the visible nodes. This doesn't account for temporary network issues,
+        // and is more of a safety check than a guarantee
+        if (reservedPool != null && generalPool != null && allAssignmentsHavePropagated(queries)) {
+            if (reservedPool.getQueries() == 0 && generalPool.getBlockedNodes() > 0) {
+                QueryExecution biggestQuery = null;
+                long maxMemory = -1;
+                for (QueryExecution queryExecution : queries) {
```

Comment:
actually never mind... the `getTotalMemoryReservation` call isn't that cheap so the fewer times we call it the better